#  EDA

In [1]:
% matplotlib inline
import pandas as pd
CS=pd.read_csv('datasets/compresive_strength_concrete.csv')

In [2]:
print(CS.head(10))

   Cement (component 1)(kg in a m^3 mixture)  \
0                                      540.0   
1                                      540.0   
2                                      332.5   
3                                      332.5   
4                                      198.6   
5                                      266.0   
6                                      380.0   
7                                      380.0   
8                                      266.0   
9                                      475.0   

   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  \
0                                                0.0       
1                                                0.0       
2                                              142.5       
3                                              142.5       
4                                              132.4       
5                                              114.0       
6                                               95.

In [3]:
print(CS.columns)

Index(['Cement (component 1)(kg in a m^3 mixture)',
       'Blast Furnace Slag (component 2)(kg in a m^3 mixture)',
       'Fly Ash (component 3)(kg in a m^3 mixture)',
       'Water  (component 4)(kg in a m^3 mixture)',
       'Superplasticizer (component 5)(kg in a m^3 mixture)',
       'Coarse Aggregate  (component 6)(kg in a m^3 mixture)',
       'Fine Aggregate (component 7)(kg in a m^3 mixture)', 'Age (day)',
       'Concrete compressive strength(MPa, megapascals) '],
      dtype='object')


It has very long column names which will be hard for us for doing further EDA as well as cleaning . So we will change column names to make it easy to  process

In [4]:
CS.columns=['Cement','Blast_Furnace_Slag','Fly_Ash','Water','Superplasticizer','Coarse_Aggregate','Fine_Aggregate','Age_day','Concrete_compressive_strength']

In [5]:
print(CS.head(10))

   Cement  Blast_Furnace_Slag  Fly_Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   
5   266.0               114.0      0.0  228.0               0.0   
6   380.0                95.0      0.0  228.0               0.0   
7   380.0                95.0      0.0  228.0               0.0   
8   266.0               114.0      0.0  228.0               0.0   
9   475.0                 0.0      0.0  228.0               0.0   

   Coarse_Aggregate  Fine_Aggregate  Age_day  Concrete_compressive_strength  
0            1040.0           676.0       28                          79.99  
1            1055.0           676.0       28                          61.89  
2             932.0         

In [6]:
print(CS.shape)

(1030, 9)


In [8]:
print(CS.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
Cement                           1030 non-null float64
Blast_Furnace_Slag               1030 non-null float64
Fly_Ash                          1030 non-null float64
Water                            1030 non-null float64
Superplasticizer                 1030 non-null float64
Coarse_Aggregate                 1030 non-null float64
Fine_Aggregate                   1030 non-null float64
Age_day                          1030 non-null int64
Concrete_compressive_strength    1030 non-null float64
dtypes: float64(8), int64(1)
memory usage: 72.5 KB
None


In [9]:
print(CS.isnull().sum())

Cement                           0
Blast_Furnace_Slag               0
Fly_Ash                          0
Water                            0
Superplasticizer                 0
Coarse_Aggregate                 0
Fine_Aggregate                   0
Age_day                          0
Concrete_compressive_strength    0
dtype: int64


In [11]:
print(CS.describe())

            Cement  Blast_Furnace_Slag      Fly_Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse_Aggregate  Fine_Aggregate      Age_day  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.

The dataset have no missing value ,  no categorical column and ranges of column are also not differ so much therefore we will not have to preprocess data

# seperating feature variables and target variable .

In [11]:
X=CS.drop('Concrete_compressive_strength',axis='columns')
Y=CS['Concrete_compressive_strength']

In [14]:
print(X.isnull().sum())

Cement                0
Blast_Furnace_Slag    0
Fly_Ash               0
Water                 0
Superplasticizer      0
Coarse_Aggregate      0
Fine_Aggregate        0
Age_day               0
dtype: int64


In [15]:
print(Y.isnull().sum())

0


All the columns of feature variable are of numerical type and no missing value .

Target variable also have no missing value . So our dataset is ready for modelling.

# LINEAR REGRESSION

In [16]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
X_train,X_test,Y_train,Y_test=train_test_split(X,Y,test_size=0.3,random_state=42)
lg=LinearRegression()
lg.fit(X_train,Y_train)
Y_pred=lg.predict(X_test)
print(r2_score(Y_test,Y_pred))


0.5943782479239204


For this dataset linear regression have R-score value is .59  which is not good.

# RIDGE REGRESSION(hyperparameter tuning)

In [17]:
from sklearn.linear_model  import Ridge
from sklearn.model_selection import GridSearchCV
param_grid={'alpha':[0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0]}
rg=Ridge()
rg_cv=GridSearchCV(rg,param_grid,cv=5)
rg_cv.fit(X_train,Y_train)
print(rg_cv.best_params_)
print(rg_cv.best_score_)

{'alpha': 1.0}
0.5965781175442143


R_square value  of Ridge regression is .596  with alpha is equal to 1.0

# DECISION TREE (Hyperparameter tuning) 

In [18]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import GridSearchCV
import numpy as np
from sklearn.metrics import r2_score
param_dt={'max_depth':np.arange(1,50),'min_samples_leaf':[0.1,0.2,0.3,0.4,0.5]}
dtr=DecisionTreeRegressor(random_state=42)
dtr_cv=GridSearchCV(dtr,param_dt,cv=10)
dtr_cv.fit(X_train,Y_train)
y_pred=dtr_cv.predict(X_test)
print(dtr_cv.best_params_)
print(r2_score(Y_test,y_pred))

{'max_depth': 5, 'min_samples_leaf': 0.1}
0.5251185899672892


R_square value of Regression decision tree is .525 with max_depth is equal to 5 and min_samples_leaf equal to 0.1

# ENSEMBLE LEARNING(BAGGING)

In [19]:
from sklearn.ensemble import BaggingRegressor
bc=BaggingRegressor(dtr_cv,n_estimators=10,n_jobs=-1,random_state=42)
bc.fit(X_train,Y_train)
pred=bc.predict(X_test)
print(r2_score(Y_test,pred))

0.6622319075977916


The R_square value of bagging classifier is 0.66 which is .14  more than the decision tree

# Random Forest Regressor

In [34]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score
Rf=RandomForestRegressor(n_estimators=400,min_samples_leaf=0.01,random_state=42)
Rf.fit(X_train,Y_train)
rf_pred=Rf.predict(X_test)
print(r2_score(Y_test,rf_pred))


0.839300899418413


Great we have now new Random_Forest_regressor model which have R2_scoreequal to 0.839 Which is good actually..... 

The best model for the given regression dataset is Random_Forest_Regresor having 400 trees with min_sample_leaf=0.01 as it have the highest R_square value = 0.839